peps/pep-0776.rst

```diff
@@ -0,0 +1,794 @@
+
+In the short term, Pyodide's packaging tooling will stay in the Pyodide
+repository. It is an open question where Pyodide's packaging tooling should live
+in the long term. Two sensible options would be for it to remain under the
+``pyodide`` organization or be moved into the ``pypa`` organization.
+
+
+--------------------------------
+Emscripten Application Lifecycle
+--------------------------------
+
+An Emscripten "binary" consists of a pair of files, an ``.mjs`` file and a
+``.wasm`` file. The ``.wasm`` file contains all of the compiled C/C++/Rust code.
+The ``.mjs`` file contains the lifecycle code to set up the runtime, locate the
+``.wasm`` file, compile it, instantiate it, call the ``main()`` function, and to
+shut down the runtime on exit. It also includes an implementation for all of the
+system calls, including the file system, the dynamic loader, and any logic to
+expose additional functionality from the JavaScript runtime to C code.
+
+The ``.mjs`` file exports a single ``bootstrapEmscriptenExecutable()``
+JavaScript function that bootstraps the runtime, calls the ``main()`` function,
+and returns an API object that can be used to call C functions. Each time it is
+called produces a complete and independent copy of the runtime with its own
+separate address space.
+
+The ``bootstrapEmscriptenExecutable()`` takes a large number of runtime
+settings. `The full list is described in the Emscripten documentation here.
+<https://emscripten.org/docs/api_reference/module.html#id3>`__ The most
+important of these are as follows:
+
+* ``thisProgram``: The value of ``argv[0]``. In Python, this makes its way into
+  ``sys.executable``.
+* ``arguments``: The list of string arguments to be passed to ``main()``.
+* ``preRun``: A list of callbacks which are invoked after the JavaScript runtime
+  and file system have been bootstrapped but before calling ``main()``. Useful
+  to set up the file system, environment variables, and standard streams.
+* ``print`` / ``printErr`` : Initial handlers for stdout and stderr. They are
+  line buffered and performing a ``flush()`` of a partial line forces an extra
+  new line. If tty-like behavior is desired, the standard stream devices should
+  be replaced in a ``preRun()`` hook.
+* ``onExit``: A handler that is called when the runtime exits.
+* ``instantiateWasm``: A callback that is called to instantiate the WebAssembly
+  module. Overriding the WebAssembly instantiation procedure via this function
+  is useful when you have other custom asynchronous startup actions or downloads
+  that can be performed in parallel to WebAssembly compilation. Implementing
+  this callback allows performing all of these in parallel.
+
+
+-----------------
+File System Setup
+-----------------
+
+In order for Python to run, it needs access to the standard library in the
+Emscripten file system. There are several possible approaches to this:
+
+* The Emscripten linker has a ``--preload-file`` flag that will automatically
```

Comment:
It exists, but if we are not using it, maybe not worth mentioning it?

Reply:
Unless it's intended to be an example of "Rejected Ideas". I don't know enough about best practices for writing PEPs..

Reply:
I'm going to leave these sections in for now. I can remove them if I get guidance that their inclusion is inappropriate.

> I don't know enough about best practices for writing PEPs..

Yeah me neither.